File: implementations/python/rabin_karp_and_rolling_hash.py
"""
Rabin-Karp and Rolling Hash implementation in Python.

Provides:
1. PrefixRollingHash with double hashing for O(1) substring hash queries.
2. RollingHashWindow for sliding window updates.
3. rabin_karp_search for exact substring matching.
4. compute_lcp for O(log n) Longest Common Prefix via binary search.
5. Comprehensive unit tests.
"""

import unittest
from typing import List, Tuple


class PrefixRollingHash:
    """Prefix rolling hash supporting O(1) substring hash queries via double hashing."""

    B1 = 911
    M1 = 1_000_000_007
    B2 = 997
    M2 = 1_000_000_009

    def __init__(self, s: str):
        self.s = s
        n = len(s)
        self.pref1 = [0] * (n + 1)
        self.pref2 = [0] * (n + 1)
        self.pow1 = [1] * (n + 1)
        self.pow2 = [1] * (n + 1)

        for i, ch in enumerate(s):
            v = ord(ch) + 1
            self.pref1[i + 1] = (self.pref1[i] * self.B1 + v) % self.M1
            self.pref2[i + 1] = (self.pref2[i] * self.B2 + v) % self.M2
            self.pow1[i + 1] = (self.pow1[i] * self.B1) % self.M1
            self.pow2[i + 1] = (self.pow2[i] * self.B2) % self.M2

    def query(self, l: int, r: int) -> Tuple[int, int]:
        """Returns double hash tuple (h1, h2) for substring s[l..r] (inclusive)."""
        if l > r or r >= len(self.s):
            raise IndexError("Substring indices out of bounds")
        length = r - l + 1
        h1 = (self.pref1[r + 1] - self.pref1[l] * self.pow1[length]) % self.M1
        h2 = (self.pref2[r + 1] - self.pref2[l] * self.pow2[length]) % self.M2
        return h1, h2


def rabin_karp_search(text: str, pattern: str) -> List[int]:
    """Finds all 0-indexed occurrences of pattern in text using Rabin-Karp."""
    n = len(text)
    m = len(pattern)
    if m == 0 or m > n:
        return []

    B = 911382323
    M = 972663749

    def val(c: str) -> int:
        return ord(c) + 1

    pattern_hash = 0
    window_hash = 0
    power = 1

    for _ in range(m - 1):
        power = (power * B) % M

    for i in range(m):
        pattern_hash = (pattern_hash * B + val(pattern[i])) % M
        window_hash = (window_hash * B + val(text[i])) % M

    matches = []
    for i in range(n - m + 1):
        if window_hash == pattern_hash:
            if text[i:i + m] == pattern:
                matches.append(i)

        if i + m < n:
            window_hash = (window_hash - val(text[i]) * power) % M
            window_hash = (window_hash * B + val(text[i + m])) % M

    return matches


def compute_lcp(h1: PrefixRollingHash, idx1: int,
                h2: PrefixRollingHash, idx2: int, max_len: int) -> int:
    """Computes length of longest common prefix of two substrings in O(log n)."""
    low, high = 1, max_len
    best = 0
    while low <= high:
        mid = (low + high) // 2
        if h1.query(idx1, idx1 + mid - 1) == h2.query(idx2, idx2 + mid - 1):
            best = mid
            low = mid + 1
        else:
            high = mid - 1
    return best


class TestRabinKarpAndRollingHash(unittest.TestCase):
    def test_rabin_karp_search(self):
        text = "AABAACAADAABAABA"
        pat = "AABA"
        self.assertEqual(rabin_karp_search(text, pat), [0, 9, 12])
        self.assertEqual(rabin_karp_search("ABCDEF", "XYZ"), [])
        self.assertEqual(rabin_karp_search("ABC", "ABC"), [0])
        self.assertEqual(rabin_karp_search("AAAA", "AA"), [0, 1, 2])
        self.assertEqual(rabin_karp_search("AB", "ABC"), [])

    def test_prefix_rolling_hash(self):
        s = "abacaba"
        prh = PrefixRollingHash(s)
        # "aba" at [0..2] and [4..6]
        self.assertEqual(prh.query(0, 2), prh.query(4, 6))
        # "aba" vs "bac"
        self.assertNotEqual(prh.query(0, 2), prh.query(1, 3))

    def test_lcp(self):
        s = "banana"
        prh = PrefixRollingHash(s)
        # Suffix "anana" (idx 1) vs "ana" (idx 3) -> LCP = 3 ("ana")
        self.assertEqual(compute_lcp(prh, 1, prh, 3, 3), 3)
        # Suffix "banana" (idx 0) vs "anana" (idx 1) -> LCP = 0
        self.assertEqual(compute_lcp(prh, 0, prh, 1, 5), 0)


if __name__ == '__main__':
    unittest.main()

File: implementations/cpp/rabin_karp_and_rolling_hash.cpp
#include <cassert>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

namespace dsa {

/**
 * @brief Double hash value pair.
 */
struct HashVal {
    uint64_t h1;
    uint64_t h2;

    bool operator==(const HashVal& o) const noexcept {
        return h1 == o.h1 && h2 == o.h2;
    }
    bool operator!=(const HashVal& o) const noexcept {
        return !(*this == o);
    }
};

/**
 * @brief Prefix rolling hash supporting O(1) substring hash queries via double hashing.
 */
class PrefixRollingHash {
private:
    static constexpr uint64_t B1 = 911ULL;
    static constexpr uint64_t M1 = 1000000007ULL;
    static constexpr uint64_t B2 = 997ULL;
    static constexpr uint64_t M2 = 1000000009ULL;

    std::string s_;
    std::vector<uint64_t> pref1_, pref2_;
    std::vector<uint64_t> pow1_, pow2_;

    static inline uint64_t char_val(char c) {
        return static_cast<uint64_t>(static_cast<unsigned char>(c)) + 1ULL;
    }

public:
    explicit PrefixRollingHash(const std::string& s)
        : s_(s), pref1_(s.size() + 1, 0), pref2_(s.size() + 1, 0),
          pow1_(s.size() + 1, 1), pow2_(s.size() + 1, 1) {
        for (size_t i = 0; i < s.size(); ++i) {
            uint64_t v = char_val(s[i]);
            pref1_[i + 1] = (pref1_[i] * B1 + v) % M1;
            pref2_[i + 1] = (pref2_[i] * B2 + v) % M2;
            pow1_[i + 1] = (pow1_[i] * B1) % M1;
            pow2_[i + 1] = (pow2_[i] * B2) % M2;
        }
    }

    size_t size() const noexcept { return s_.size(); }

    /**
     * @brief Returns the double hash of substring s[l..r] (inclusive).
     */
    HashVal query(size_t l, size_t r) const {
        if (l > r || r >= s_.size()) {
            throw std::out_of_range("PrefixRollingHash::query index out of range");
        }
        size_t len = r - l + 1;

        int64_t val1 = static_cast<int64_t>(pref1_[r + 1]) -
                       static_cast<int64_t>((pref1_[l] * pow1_[len]) % M1);
        if (val1 < 0) val1 += M1;

        int64_t val2 = static_cast<int64_t>(pref2_[r + 1]) -
                       static_cast<int64_t>((pref2_[l] * pow2_[len]) % M2);
        if (val2 < 0) val2 += M2;

        return {static_cast<uint64_t>(val1), static_cast<uint64_t>(val2)};
    }
};

/**
 * @brief Rabin-Karp exact string matching algorithm.
 * Returns all starting indices where pattern occurs in text.
 */
inline std::vector<int> rabin_karp_search(const std::string& text, const std::string& pattern) {
    std::vector<int> matches;
    int n = static_cast<int>(text.size());
    int m = static_cast<int>(pattern.size());
    if (m == 0 || m > n) return matches;

    const int64_t B = 911382323LL;
    const int64_t M = 972663749LL;

    auto val = [](char c) -> int64_t {
        return static_cast<int64_t>(static_cast<unsigned char>(c)) + 1LL;
    };

    int64_t pattern_hash = 0;
    int64_t window_hash = 0;
    int64_t power = 1;

    for (int i = 0; i < m - 1; ++i) {
        power = (power * B) % M;
    }

    for (int i = 0; i < m; ++i) {
        pattern_hash = (pattern_hash * B + val(pattern[i])) % M;
        window_hash = (window_hash * B + val(text[i])) % M;
    }

    for (int i = 0; i + m <= n; ++i) {
        if (window_hash == pattern_hash) {
            // Confirm with direct character comparison to eliminate collisions
            if (text.compare(i, m, pattern) == 0) {
                matches.push_back(i);
            }
        }

        if (i + m < n) {
            window_hash = (window_hash - val(text[i]) * power) % M;
            if (window_hash < 0) window_hash += M;
            window_hash = (window_hash * B + val(text[i + m])) % M;
        }
    }

    return matches;
}

/**
 * @brief Compute Longest Common Prefix (LCP) of two substrings using binary search and prefix hashing.
 * Time complexity: O(log(min(len1, len2))).
 */
inline size_t compute_lcp(const PrefixRollingHash& h1, size_t idx1,
                          const PrefixRollingHash& h2, size_t idx2,
                          size_t max_len) {
    size_t low = 0, high = max_len;
    size_t best = 0;

    while (low <= high) {
        size_t mid = low + (high - low) / 2;
        if (mid == 0) {
            low = 1;
            continue;
        }
        if (h1.query(idx1, idx1 + mid - 1) == h2.query(idx2, idx2 + mid - 1)) {
            best = mid;
            low = mid + 1; // Try longer prefix
        } else {
            high = mid - 1;
        }
    }
    return best;
}

} // namespace dsa

int main() {
    using namespace dsa;

    std::cout << "[RUNNING] Rabin-Karp and Rolling Hash C++17 Verification..." << std::endl;

    // 1. Rabin-Karp Search
    {
        std::string text = "AABAACAADAABAABA";
        std::string pat = "AABA";
        auto matches = rabin_karp_search(text, pat);
        std::vector<int> expected = {0, 9, 12};
        assert(matches == expected);

        // Pattern not present
        assert(rabin_karp_search("ABCDEF", "XYZ").empty());

        // Pattern equals text
        assert(rabin_karp_search("ABC", "ABC") == std::vector<int>{0});

        // Pattern longer than text
        assert(rabin_karp_search("AB", "ABC").empty());

        // Overlapping matches
        std::string text2 = "AAAA";
        std::string pat2 = "AA";
        assert((rabin_karp_search(text2, pat2) == std::vector<int>{0, 1, 2}));
    }

    // 2. Prefix Rolling Hash Substring Queries
    {
        std::string s = "abacaba";
        PrefixRollingHash prh(s);

        // Identical substrings "aba" at indices [0..2] and [4..6]
        HashVal h1 = prh.query(0, 2);
        HashVal h2 = prh.query(4, 6);
        assert(h1 == h2);

        // Distinct substrings "aba" vs "bac"
        HashVal h3 = prh.query(1, 3);
        assert(h1 != h3);
    }

    // 3. Binary Search LCP with Prefix Rolling Hash
    {
        std::string s = "banana";
        PrefixRollingHash prh(s);

        // Suffix "anana" (index 1) vs "ana" (index 3)
        // LCP should be "ana" (length 3)
        size_t lcp = compute_lcp(prh, 1, prh, 3, 3);
        assert(lcp == 3);

        // Suffix "banana" (index 0) vs "anana" (index 1) -> LCP = 0
        size_t lcp2 = compute_lcp(prh, 0, prh, 1, 5);
        assert(lcp2 == 0);
    }

    std::cout << "[PASSED] Rabin-Karp and Rolling Hash C++17 All Tests Passed!" << std::endl;
    return 0;
}
